twill-yarn/src/main/java/org/apache/twill/internal/appmaster/ApplicationMasterService.java

```diff
@@ -181,19 +190,130 @@ public Reader getInput() throws IOException {
   }
 
   @SuppressWarnings("unchecked")
-  @Nullable
   private EventHandler createEventHandler(TwillSpecification twillSpec) throws ClassNotFoundException {
     // Should be able to load by this class ClassLoader, as they packaged in the same jar.
     EventHandlerSpecification handlerSpec = twillSpec.getEventHandler();
     if (handlerSpec == null) {
-      return null;
+      // if no handler is specified, return an EventHandler with no-op
+      return new EventHandler() {};
     }
 
     Class<?> handlerClass = getClass().getClassLoader().loadClass(handlerSpec.getClassName());
     Preconditions.checkArgument(EventHandler.class.isAssignableFrom(handlerClass),
                                 "Class {} does not implements {}",
                                 handlerClass, EventHandler.class.getName());
-    return Instances.newInstance((Class<? extends EventHandler>) handlerClass);
+    final EventHandler delegate = Instances.newInstance((Class<? extends EventHandler>) handlerClass);
+    // wrap all calls to the delegate EventHandler methods except initialize so that all errors will be caught
+    return new EventHandler() {
+      private boolean initialized;
+
+      @Override
+      public void initialize(EventHandlerContext context) {
+        delegate.initialize(context);
+        initialized = true;
```

Comment:
Since any exception thrown from `initialize` will terminate the app, I think we don't need this, right?